PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+#pragma once
+
+#include "BTreeNode.hpp"
+#include "Optional.hpp"
+
+namespace Poly {
+
+	template<typename K, typename V, size_t B>
+	struct Range { //note(vuko): doubles as a proper fucking iterator; behold the goddamn beauty! (alright, this is for show only for now)
+		using BTree  = Impl::BTree<K, V, B>;
+		using KVERef = typename BTree::KVERef;
+
+		KVERef front;
+		KVERef back;
+
+		struct KV {
+			const K& key;
+			V& value;
+		};
+
+		KV next_unchecked() {
+			KVERef kv_ref = this->front;
+
+			if (kv_ref.idx < kv_ref.node_ref.node->len) {
+				front = KVERef{kv_ref.node_ref, kv_ref.idx + 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx >= kv_ref.node_ref.node->len);
+
+			front = BTree::first_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx + 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV prev_unchecked() {
+			KVERef kv_ref = this->back;
+
+			if (kv_ref.idx > 0) {
+				this->back = KVERef{kv_ref.node_ref, kv_ref.idx - 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx == 0);
+
+			this->back = BTree::last_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx - 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV next() {
+			assert(!(front == back));
+			return next_unchecked();
+		}
+	};
+
+	//todo(vuko): docs
+	template<typename K, typename V, size_t Bfactor = 6>
+	class BTreeMap {
+		static_assert(Bfactor > 1, "B factor must be greater than 1. Consider using a classic binary tree if you need a lesser value.");
+		using BTree    = Impl::BTree<K, V, Bfactor>;
+		using Root     = typename BTree::Root;
+		using LeafNode = typename BTree::LeafNode;
+		using KVERef   = typename BTree::KVERef;
+		using NodeRef  = typename BTree::NodeRef;
+		class Entry;
+		class Iter;
+		class Keys;
+		class Values;
+	public:
+		static constexpr size_t B = Bfactor;
+
+	public:
+		BTreeMap() : root{new LeafNode(), 0}, len(0) {} //todo(vuko): should we delay the allocation until the first insertion?
+		BTreeMap(BTreeMap&& other) : root(other.root), len(other.len) { ::new(&other) BTreeMap(); } //note(vuko): due to the previous issue this allocates!
+		//todo(vuko): copy constructor? tbh I'd rather avoid implementing it
+		~BTreeMap() { this->Clear(); delete this->root.node; };
+
+	public:
+		BTreeMap& operator=(BTreeMap&& other) {
+			this->~BTreeMap();
+			this->root = other.root;
+			this->len = other.len;
+			::new(&other) BTreeMap();
+			return *this;
+		}
+
+	public:
+		Optional<V> Insert(const K&  key, const V&  value) { return this->insert(          key ,           value ); }
```

Comment:
What is wrong with `this->`? (pun not intended)

>rename to InsertImpl

Pimples, yuck...
I agree that the public API needs to be consistent and feel that lowercase communicates that the inspected method is not a part it. If that's not acceptable, how
about `ForwardingInsert`?

>What is in the returned Optional<V>?

`Insert()` returns an empty Optional in case the key was not present in the map. Otherwise, the value is updated and the old value is returned. Note that the key is not updated (may matter for types, that can be equal without being identical).

>This whole api needs documentation

I want to write one, once the API is stabilized. In the meantime, one can infer most notions about intended usage and the API details from the unit tests code ;)